server/src/main/java/io/crate/execution/ddl/tables/AlterTableClient.java

```diff
@@ -237,44 +235,30 @@ private CompletableFuture<Long> resize(BoundAlterTable analysis) {
         validateNumberOfShardsForResize(sourceIndexMetadata, targetNumberOfShards);
         validateReadOnlyIndexForResize(sourceIndexMetadata);
 
-        final List<ChainableAction<Long>> actions = new ArrayList<>();
         final String resizedIndex = RESIZE_PREFIX + sourceIndexName;
-        deleteLeftOverFromPreviousOperations(currentState, actions, resizedIndex);
+
+        CompletableFuture<Long> future;
+        if (currentState.metadata().hasIndex(resizedIndex)) {
+            future = deleteIndex(resizedIndex);
+        } else {
+            future = CompletableFuture.completedFuture(1L);
+        }
 
         final ResizeRequest request = new ResizeRequest(
             table.ident(),
             partitionName == null ? List.of() : partitionName.values(),
             targetNumberOfShards
         );
-
-        actions.add(new ChainableAction<>(
-            () -> transportResizeAction.execute(request, r -> r.isAcknowledged() ? 1L : 0L),
-            () -> CompletableFuture.completedFuture(-1L)
-        ));
-        actions.add(new ChainableAction<>(
-            () -> swapAndDropIndex(resizedIndex, sourceIndexName)
-                .exceptionally(error -> {
+        return future.thenCompose(_ -> transportResizeAction.execute(request).thenCompose(response -> {
+            if (response.isAcknowledged() && response.isShardsAcknowledged()) {
+                return swapAndDropIndex(resizedIndex, sourceIndexName);
+            } else {
+                return deleteIndex(resizedIndex).handle((_, _) -> {
                     throw new IllegalStateException(
-                        "The resize operation to change the number of shards completed partially but run into a failure. " +
-                        "Please retry the operation or clean up the internal indices with ALTER CLUSTER GC DANGLING ARTIFACTS. "
-                        + error.getMessage(), error
-                    );
-                }),
-            () -> CompletableFuture.completedFuture(-1L)
-        ));
-        return ChainableActions.run(actions);
-    }
-
-    private void deleteLeftOverFromPreviousOperations(ClusterState currentState,
-                                                      List<ChainableAction<Long>> actions,
-                                                      String resizeIndex) {
-
-        if (currentState.metadata().hasIndex(resizeIndex)) {
-            actions.add(new ChainableAction<>(
-                () -> deleteIndex(resizeIndex),
-                () -> CompletableFuture.completedFuture(-1L)
-            ));
-        }
+                        "Resize operation wasn't acknowledged. Check shard allocation and retry");
```

Comment:
shouldn't we include the original exception, which prevents the operation?

Reply:
In most cases there won't be an original exception because the resize itself doesn't fail directly, it succeeds but without the shards getting allocated in time.

The `deleteIndex` is also only to cleanup - and likely doesn't fail - if it would it is also not the original error.
But I can add it as cause if there is one anyway

Reply:
Thx for explaining, then I'll leave it up to you if it worths including the original exception here in the handling.